torch/ao/sparsity/experimental/activation_sparsifier/activation_sparsifier.py

```diff
@@ -270,10 +270,34 @@ def update_mask(self, name, data, configs):
                 data_feature = reduce_fn(data[feature_idx])
                 mask[feature_idx].data = mask_fn(data_feature, **sparse_config)
 
+    def __sparsify_hook(self, name):
+        """Returns hook that applies sparsification mask to input entering the attached layer
+        """
+        mask = self.get_mask(name)
+        features = self.data_groups[name]['features']
+        feature_dim = self.data_groups[name]['feature_dim']
+
+        def hook(module, input):
+            input_data = input[0]
+            if features is None:
+                # apply to all the features
+                return input_data * mask
+            else:
+                # apply per feature, feature_dim
+                for feature_idx in range(0, len(features)):
+                    feature = torch.Tensor([features[feature_idx]], device=input_data.device).long()
+                    sparsified = torch.index_select(input_data, feature_dim, feature) * mask[feature_idx]
+                    input_data.index_copy_(feature_dim, feature, sparsified)
+                return input_data
+        return hook
+
     def squash_mask(self, **kwargs):
```

Comment:
Aren't you supposed to deregister the hooks at this point? Does `.remove` do that?

Also, I think we should ask the user if they want to keep the forward pre hook. I am just thinking, if the user just want the masks, then the squash_mask is supposed to compute it, and get out, right? So a user can *optionally* request the forward hook to stay

Reply:
yes ```.remove()``` deregisters the hook as mentioned in [docs](https://pytorch.org/docs/stable/generated/torch.nn.modules.module.register_module_forward_pre_hook.html)

Also, ```squash_mask()``` does not compute the mask, ```update_mask()``` does. We can add an argument ```attach_sparsify_hook``` which if True attache the sparsify forward pre-hook, otherwise just removes the aggregate handle